### PR TITLE
[compatible] Track nodes going offline in integration tests, fast fail if required

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds a mechanism for tracking nodes that go offline during unit tests.

This then modifies each test so that
* it logs any time a node goes offline
* if any nodes are 'required to be online' by the test, we halt the test immediately (by propagating an exception)

The goal of this is to get better visibility into node restarts, as well as to prevent 'zombie' tests that will fail later during their execution.

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

File: src/app/test_executive/delegation_test.ml [hunk 1]
```diff
@@ -44,6 +44,17 @@ module Make (Inputs : Intf.Test.Inputs_intf) = struct
...
+
+    let online_monitor, online_monitor_subscription =
+      Wait_condition.monitor_online_nodes ~logger (event_router t)
+    in
+    (* The node that we send GraphQL requests to needs to stay online. *)
+    Wait_condition.require_online online_monitor node_b ;
+    (* The archive node needs to be online to run the final replayer test. *)
+    Core_kernel.Map.iter
+      ~f:(Wait_condition.require_online online_monitor)
+      (Network.archive_nodes network) ;
```
Comment (deepthiskumar) [inline src/app/test_executive/delegation_test.ml:56]: I don't see replayer used in this test. Is archive node used elsewhere?

File: src/app/test_executive/medium_bootstrap.ml [hunk 2]
```diff
@@ -61,6 +61,14 @@ module Make (Inputs : Intf.Test.Inputs_intf) = struct
     let node_c =
       Core.String.Map.find_exn (Network.block_producers network) "node-c"
     in
+
+    let online_monitor, online_monitor_subscription =
+      Wait_condition.monitor_online_nodes ~logger (event_router t)
+    in
+    (* The nodes that we synchronize to must be online. *)
+    Wait_condition.require_online online_monitor node_a ;
+    Wait_condition.require_online online_monitor node_b ;
```
Comment (deepthiskumar) [inline src/app/test_executive/medium_bootstrap.ml:70]: should `node_c` be made required-online after it restarts in the section below?

File: src/app/test_executive/payments_test.ml [hunk 3]
```diff
@@ -85,6 +85,17 @@ module Make (Inputs : Intf.Test.Inputs_intf) = struct
...
+
+    let online_monitor, online_monitor_subscription =
+      Wait_condition.monitor_online_nodes ~logger (event_router t)
+    in
+    (* The node that we send GraphQL requests to needs to stay online. *)
+    Map.iter ~f:(Wait_condition.require_online online_monitor) all_nodes ;
+    (* We need snark work for this test. *)
+    Core_kernel.Map.iter
+      ~f:(Wait_condition.require_online online_monitor)
+      (Network.snark_coordinators network) ;
```
Comment (deepthiskumar) [inline src/app/test_executive/payments_test.ml:97]: `all_nodes` includes snark coordinator as well

File: src/app/test_executive/peers_reliability_test.ml [hunk 4]
```diff
@@ -57,6 +57,14 @@ module Make (Inputs : Intf.Test.Inputs_intf) = struct
     let node_c =
       Core.String.Map.find_exn (Network.block_producers network) "node-c"
     in
+
+    let online_monitor, online_monitor_subscription =
+      Wait_condition.monitor_online_nodes ~logger (event_router t)
+    in
+    (* The nodes that we synchronize to must be online. *)
+    Wait_condition.require_online online_monitor node_a ;
+    Wait_condition.require_online online_monitor node_b ;
```
Comment (deepthiskumar) [inline src/app/test_executive/peers_reliability_test.ml:66]: should `node_c` be made required-online after it restarts in the section below?

File: src/app/test_executive/chain_reliability_test.ml [hunk 5]
```diff
@@ -49,6 +49,14 @@ module Make (Inputs : Intf.Test.Inputs_intf) = struct
     let node_c =
       Core.String.Map.find_exn (Network.block_producers network) "node-c"
     in
+
+    let online_monitor, online_monitor_subscription =
+      Wait_condition.monitor_online_nodes ~logger (event_router t)
+    in
+    (* The nodes that we sync to need to stay online. *)
+    Wait_condition.require_online online_monitor node_a ;
+    Wait_condition.require_online online_monitor node_b ;
```
Comment (deepthiskumar) [inline src/app/test_executive/chain_reliability_test.ml:58]: should `node_c` be made required-online after it restarts in the section below?

**Comment by mrmr1993** [comment]
!ci-build-me

**Review by nholland94 — APPROVED** [review]

**Pull request merged**